packages/sdk/src/bot/client/types.test.ts

```diff
@@ -0,0 +1,291 @@
+      messages: {
+        messageBaz: {
+          baz: boolean
+        }
+      }
+    }>
+  }
+}>
+
+test('EnumerateActions should enumerate actions', () => {
+  type Bot = MakeBot<{
+    integrations: {
+      fooBarBaz: FooBarBazIntegration
+    }
+  }>
+  type Actual = types.EnumerateActions<Bot>
+  type Expected = {
+    'fooBarBaz:doFoo': FooBarBazIntegration['actions']['doFoo']
+    'fooBarBaz:doBar': FooBarBazIntegration['actions']['doBar']
+    'fooBarBaz:doBaz': FooBarBazIntegration['actions']['doBaz']
+  }
+  type _assertion = utils.AssertAll<
+    [
+      //
+      utils.AssertExtends<Actual, Expected>,
+      utils.AssertExtends<Expected, Actual>
+    ]
+  >
+})
+
+test('EnumerateActions should return empty object if TBot is BaseBot', () => {
+  type Actual = types.EnumerateActions<BaseBot>
+  type Expected = {}
+  type _assertion = utils.AssertAll<
+    [
+      //
+      utils.AssertExtends<keyof Actual, keyof Expected>,
+      utils.AssertExtends<keyof Expected, keyof Actual>
+    ]
+  >
+})
+
+test('EnumerateEvents should enumerate events', () => {
+  type Bot = MakeBot<{
+    integrations: {
+      fooBarBaz: FooBarBazIntegration
+    }
+    events: {
+      onQux: {
+        eventQux: null
+      }
+    }
+  }>
+  type Actual = types.EnumerateEvents<Bot>
+  type Expected = {
+    'fooBarBaz:onFoo': FooBarBazIntegration['events']['onFoo']
+    'fooBarBaz:onBar': FooBarBazIntegration['events']['onBar']
+    'fooBarBaz:onBaz': FooBarBazIntegration['events']['onBaz']
+    onQux: Bot['events']['onQux']
+  }
+  type _assertion = utils.AssertAll<
+    [
+      //
+      utils.AssertExtends<Actual, Expected>,
+      utils.AssertExtends<Expected, Actual>
+    ]
+  >
+})
+
+test('EnumerateEvents with only integration events should enumerate only integraiton events', () => {
+  type Bot = MakeBot<{
+    integrations: {
+      fooBarBaz: FooBarBazIntegration
+    }
+  }>
+  type Actual = types.EnumerateEvents<Bot>
+  type Expected = {
+    'fooBarBaz:onFoo': FooBarBazIntegration['events']['onFoo']
+    'fooBarBaz:onBar': FooBarBazIntegration['events']['onBar']
+    'fooBarBaz:onBaz': FooBarBazIntegration['events']['onBaz']
+  }
+  type _assertion = utils.AssertAll<
+    [
+      //
+      utils.AssertExtends<Actual, Expected>,
+      utils.AssertExtends<Expected, Actual>
+    ]
+  >
+})
+
+test('EnumerateEvents with only bot events should enumerate only bot events', () => {
+  type Bot = MakeBot<{
+    events: {
+      onQux: {
+        eventQux: null
+      }
+    }
+  }>
+  type Actual = types.EnumerateEvents<Bot>
+  type Expected = {
+    onQux: Bot['events']['onQux']
+  }
+  type _assertion = utils.AssertAll<
+    [
+      //
+      utils.AssertExtends<Actual, Expected>,
+      utils.AssertExtends<Expected, Actual>
+    ]
+  >
+})
+
+test('EnumerateEvents should return empty object if TBot is BaseBot', () => {
+  type Actual = types.EnumerateEvents<BaseBot>
+  type Expected = {}
+  type _assertion = utils.AssertAll<
```

Comment:
**question:** what is the benefit of having an `AssertAll` utility type? wouldn't `AssertAll<[ A, B, C ]>` just throw at the very first error and not execute the rest?

Reply:
honestly, I don't know. It was to prevent writing:
```ts
type _assertion1 = utils.AssertExtends<Actual, Expected>
type _assertion2 = utils.AssertExtends<Expected, Actual>
```
Also, the following kind of sucks:
```ts
type _assertion = utils.AssertTrue<utils.IsEqual<Actual, Expected>>
```
As the only error message you'll get is that true is not false